File: prob2.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if(len(nums)==0):
            return []
        nums.sort()
        res = []
        
        print (nums)
        n = len(nums)
        for i in range(n):
            if(i > 0 and nums[i] == nums[i-1]):
                continue
            l = i + 1
            r = n -1
            r=len(nums)-1
            while(l<r):
                sum = nums[i]+nums[l]+nums[r]
                if (sum <0):
                    l = l+1
                
                elif(sum>0):
                    r =r-1    
                else:
                    res.append([nums[i],nums[l],nums[r]])
                    l += 1
                    r -= 1
                    while(l < r and nums[l]==nums[l-1]):
                        l += 1
                    while(l < r and nums[r]==nums[r+1]):
                        r -= 1
        
        return res
            

File: prob2.java
// Time Complexity : O (n^2) 
// Space Complexity : O (1) no extra space
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Minor issues, sorting the array, checking base condition inside while loop

import java.util.*;
public class prob2 {

    private static List<List<Integer>> threeSum(int[] nums){
        if(nums == null || nums.length == 0) return new ArrayList<>();
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums); // can apply 2 ptr technique only on sorted array O(nlogn)

        // O(n^2)
        for(int i = 0; i < nums.length; i++){
            if(i > 0 && nums[i] == nums[i - 1]) continue; // skip duplicate elements in outer loop
            int lo = i+1; int hi = nums.length - 1;
            while(lo < hi){
                int sum = nums[i] + nums[lo] + nums[hi];
                if(sum == 0){
                    List<Integer> li = Arrays.asList(nums[i], nums[lo], nums[hi]);
                    result.add(li);
                    lo++; hi--;
                    while(lo < hi && nums[lo] == nums[lo-1])lo++; // check base condition here also and skip dupes
                    while(lo < hi && nums[hi] == nums[hi+1])hi--;

                }else if(sum < 0) lo++;
                else hi--;
            }
        }
        return result;
    }
    public static void main(String[] args) {
        int[] nums = {-1,0,1,2,-1,-4};
        System.out.println(threeSum(nums));
    }
    
}
